File: src/main/kotlin/easy/SameTree.kt
package easy

import datastructures.TreeNode

object SameTree {

    fun isSameTree(p: TreeNode?, q: TreeNode?): Boolean {
        return recursion(p, q)
    }

    fun recursion(p: TreeNode?, q: TreeNode?): Boolean {
        if(p == null && q == null) return true
        if(p == null || q == null) return false

        if(p.`val` != q.`val`) return false

        return recursion(p.left, q.left) && recursion(p.right, q.right)
    }
}

File: leetcode/easy/same_tree.go
package easy

import . "github.com/johnazedo/interview-questions/leetcode"

/*
Number: 100
Difficult: Easy
Link: https://leetcode.com/problems/same-tree/
Tags: Binary Tree, Depth-First Search
Status: Resolved
*/
func isSameTree(p *TreeNode, q *TreeNode) bool {
	// TODO: Do this iteratively
	// Time: O(N)
	// Space: O(H)
	if p == nil && q == nil {
		return true
	}
	if p == nil || q == nil {
		return false
	}

	return p.Val == q.Val && isSameTree(p.Left, q.Left) && isSameTree(p.Right, q.Right)
}
